File: Ch02/Ex2_03.hs
{-
 - Ex2_02.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Represent rectangles on a plane
 -
 -}

module Ex2_03 where
import Ex2_02

newtype Rectangle = MakeRectangle (Point, Point, Point, Point)

toRectangle :: (Point, Point, Point, Point) -> Rectangle
toRectangle (p1,p2,p3,p4) = MakeRectangle (p1,p2,p3,p4)

fromRectangle :: Rectangle -> (Point, Point, Point, Point)
fromRectangle (MakeRectangle (p1,p2,p3,p4)) = (p1,p2,p3,p4)

upperLeft :: Rectangle -> Point
upperLeft s = ul
  where (ul, _, _, _) = fromRectangle s

upperRight :: Rectangle -> Point
upperRight s = ur
  where (_, ur, _, _) = fromRectangle s

lowerLeft :: Rectangle -> Point
lowerLeft s = ll
  where (_, _, ll, _) = fromRectangle s

lowerRight :: Rectangle -> Point
lowerRight s = lr
  where (_, _, _, lr) = fromRectangle s

perimeter :: Rectangle -> Float
perimeter s = (top + side) * 2
  where top             = distance ul ur
        side            = distance ur lr
        distance p1 p2  = sqrt (square(xDiff p1 p2) + square(yDiff p1 p2))
        xDiff p1 p2     = (xPoint p2) - (xPoint p1)
        yDiff p1 p2     = (yPoint p2) - (yPoint p1)
        ul              = upperLeft s
        ur              = upperRight s
        lr              = lowerRight s
        square x       = x * x

area :: Rectangle -> Float
area s = top * side
  where top             = distance ul ur
        side            = distance ur lr
        distance p1 p2  = sqrt (square(xDiff p1 p2) + square(yDiff p1 p2))
        xDiff p1 p2     = (xPoint p2) - (xPoint p1)
        yDiff p1 p2     = (yPoint p2) - (yPoint p1)
        ul              = upperLeft s
        ur              = upperRight s
        lr              = lowerRight s
        square x       = x * x

File: Ch02/ex2_03.erl
-module(ex2_03).
-author("Kevin C. Baird").
-purpose("SICP in Erlang: Exercise 2.3 - Represent rectangle on a plane.").
-export([make_rectangle/4, perimeter/1, area/1]).
-import(ex2_03, [make_point/2, x_point/1, y_point/1]).
-record(point,     {x,y}).
-record(rectangle, {ul,ur,ll,lr}).

% creators
make_rectangle({ul, X1, Y1}, {ur, X2, Y2}, {ll, X3, Y3}, {lr, X4, Y4}) ->
  #rectangle{ul=make_point(X1,Y1),ur=make_point(X2,Y2),ll=make_point(X3,Y3),lr=make_point(X4,Y4)}.

% extractors
area(R) ->
  (distance(R#rectangle.ul, R#rectangle.ur) *
   distance(R#rectangle.ur, R#rectangle.lr)).

perimeter(R) ->
  (distance(R#rectangle.ul, R#rectangle.ur) +
   distance(R#rectangle.ur, R#rectangle.lr)) * 2.0.

% hidden helpers
distance(P1, P2) ->
  math:sqrt(square(x_diff(P1, P2)) + square(y_diff(P1, P2))).
square(X) -> X * X.
x_diff(P1, P2) -> P2#point.x - P1#point.x.
y_diff(P1, P2) -> P2#point.y - P1#point.y.

